FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/fragment/dialog/FixNeedPassphraseIssueDialogFragment.kt

```diff
@@ -274,6 +299,26 @@ class FixNeedPassphraseIssueDialogFragment : BaseDialogFragment() {
         else -> {}
       }
     }
+
+    launchAndRepeatWithLifecycle {
+      checkPrivateKeysViewModel.antiBruteforceProtectionCountdownStateFlow.collect {
+        val attemptsCount = it.first
+        val timeLeft = it.second
+        binding?.btnUpdatePassphrase?.isEnabled = timeLeft == 0L
+        if (timeLeft == 0L) {
+          if (attemptsCount == 0) {
+            binding?.tILKeyPassword?.error = null
+          }
+          binding?.btnUpdatePassphrase?.text = getString(R.string.provide_passphrase)
+        } else {
+          binding?.tILKeyPassword?.error = getString(
+            R.string.private_key_passphrase_anti_bruteforce_protection_hint
+          )
+          binding?.btnUpdatePassphrase?.text =
+            SimpleDateFormat("mm:ss", Locale.getDefault()).format(Date(timeLeft))
+        }
+      }
+    }
```

Comment:
We may simplify as belows. 
- use attempsCount, timeLeft as tuple
- use `val binding = binding ?: return@collect` instead of using `binding?.` everytime
- use `when` instead of if
- simplify `btnUpdatePassphrase` set text logic

```kt
checkPrivateKeysViewModel.antiBruteforceProtectionCountdownStateFlow.collect { (attemptsCount, timeLeft) ->
    val binding = binding ?: return@collect

    binding.btnUpdatePassphrase.isEnabled = timeLeft == 0L
    binding.btnUpdatePassphrase.text = when {
        timeLeft > 0L -> {
            binding.tILKeyPassword.error = getString(R.string.private_key_passphrase_anti_bruteforce_protection_hint)
            SimpleDateFormat("mm:ss", Locale.getDefault()).format(Date(timeLeft))
        }
        else -> {
            if (attemptsCount == 0) binding.tILKeyPassword.error = null
            getString(R.string.provide_passphrase)
        }
    }
}

```

Reply:
1. I've applied `use attempsCount, timeLeft as tuple`. This simplification looks great.
2. `use val binding = binding ?: return@collect instead of using binding?. everytime` - not sure about this one. `?` was created for [null-safety](https://kotlinlang.org/docs/null-safety.html#safe-calls). I prefer to leave it as-is.
3. `simplify btnUpdatePassphrase set text logic`. I prefer to leave it as-is. IMHO, it's not a good idea to do some actions when it doesn't relate to things that you were going to do.

How I see (and read it):

`binding.btnUpdatePassphrase.text = when {` - means we do some action depending on a situation and return some result to `text`. As for me, in this situation in `when` should not be any additional operations.

I agree with  

```
binding.btnUpdatePassphrase.text = when {
        timeLeft > 0L -> {
            SimpleDateFormat("mm:ss", Locale.getDefault()).format(Date(timeLeft))
        }
        else -> {
            getString(R.string.provide_passphrase)
        }
    }
```

but not with

```
binding.btnUpdatePassphrase.text = when {
        timeLeft > 0L -> {
            //additional action start
            binding.tILKeyPassword.error = getString(R.string.private_key_passphrase_anti_bruteforce_protection_hint)
            //additional action stop
            SimpleDateFormat("mm:ss", Locale.getDefault()).format(Date(timeLeft))
        }
        else -> {
            //additional action start
            if (attemptsCount == 0) binding.tILKeyPassword.error = null
            //additional action stop
            getString(R.string.provide_passphrase)
        }
    }
```

Reply:
Yeah, it's for null-safety and that's why we have `return@collect` in case binding is null.
 
> use val binding = binding ?: return@collect instead of using binding?. everytime - not sure about this one. ? was created for [null-safety](https://kotlinlang.org/docs/null-safety.html#safe-calls). I prefer to leave it as-is.

Yeah, I agree. I tried to simplify too much and below looks good.
>  I agree with

> binding.btnUpdatePassphrase.text = when {
        timeLeft > 0L -> {
            SimpleDateFormat("mm:ss", Locale.getDefault()).format(Date(timeLeft))
        }
        else -> {
            getString(R.string.provide_passphrase)
        }
    }

Reply:
For the second point, also in other languages like JavaScript/Swift, when dealing with optional values multiple times, I prefer to first check if the value is null and then return/break accordingly.

like in javascript
```js
if (!value) { return }
// proceed with non-null value
```
in swift
```swift
guard let value = value else { return }
// proceed with non-null value
```

Reply:
For 3rd point, we may make code like below
```kt
when {
  timeLeft > 0L -> {
    binding.tILKeyPassword.error = getString(R.string.private_key_passphrase_anti_bruteforce_protection_hint)
    binding.btnUpdatePassphrase.text = SimpleDateFormat("mm:ss", Locale.getDefault()).format(Date(timeLeft))
  }
  else -> {
    if (attemptsCount == 0) binding.tILKeyPassword.error = null
    binding.btnUpdatePassphrase.text = getString(R.string.provide_passphrase)
  }
}
```

Reply:
But I think it's better to add just one line of code (convert/set optional value to non-nullable value)  instead of adding `?` when you have to deal with this value?

Reply:
> For 3rd point, we may make code like below

Agree. But I don't see a lot of simplification for a situation where we have only 2 cases. `If/else` can be used here. 
Also, we prefer to use 

```
if (attemptsCount == 0) {
            binding?.tILKeyPassword?.error = null
}
```

instead of

`if (attemptsCount == 0) binding.tILKeyPassword.error = null`

Sorry, I can't find the conversation about that(it was 1-2 years ago, maybe more). As I remember we've decided to do that because using `{}` looks better for reviewing for outside reviewers.

Reply:
I agree.
> But I don't see a lot of simplification for a situation where we have only 2 cases. If/else can be used here.

Reply:
> instead of adding ? multiple times when you have to deal with this value?

I'm totally comfortable with this :)). But if we are talking about this particular situation. Other cases should use the logic that you have proposed.

Reply:
> I'm totally comfortable with this :)).

Just in case. It's a common approach in Android(Kotlin).